File: test/bench/dbtest_neo4j.py
import time
from neo4j import GraphDatabase

"""

Random write: 245.33559386034676 objects/second
Bulk read: 52825.918844916654 objects/second
Random read: 223.60659061924466 objects/second

"""

db = GraphDatabase.driver("bolt://localhost", auth = ("", ""))
session = db.session()

NB = 10000

"""
Supprime toute la base :

MATCH (n)  OPTIONAL MATCH (n)-[r]-()  DELETE n,r

"""
    
t = time.time()
#tx = session.begin_transaction()
#tx.run("""CREATE INDEX ON :Node(label)""")
#tx.commit()
tx = session.begin_transaction()
previous = None
for i in range(NB):
  label = "node number %s" % i
  node = tx.run("""CREATE (x:Node {label:"%s"}) RETURN id(x)""" % label).value()[0]
  if previous:
    tx.run("""MATCH (x:Node) WHERE id(x)=%s MATCH (y:Node) WHERE id(y)=%s CREATE (x)-[:NEXT]->(y)""" % (previous, node))
  previous = node
tx.commit()
dt = time.time() - t
print("Random write:", 1/dt*NB, "objects/second")


t = time.time()
tx = session.begin_transaction()
labels = tx.run("""MATCH (x:Node) RETURN x.label""").value()
nexts  = tx.run("""MATCH (x:Node)-[:NEXT]->(y:Node) RETURN id(x), id(y)""").values()
tx.commit()
dt = time.time() - t
print("Bulk read:", 1/dt*NB, "objects/second")

t = time.time()
tx = session.begin_transaction()
node = tx.run("""MATCH (x:Node) WHERE x.label="node number 0" RETURN id(x)""").value()[0]
while True:
  label = tx.run("""MATCH (x:Node) WHERE id(x)=%s RETURN x.label""" % node).value()[0]
  nexts = tx.run("""MATCH (x:Node)-[:NEXT]->(y:Node) WHERE id(x)=%s RETURN id(y)""" % node).value()
  if nexts:
    node = nexts[0]
  else:
    break
dt = time.time() - t
print("Random read:", 1/dt*NB, "objects/second")

File: test/bench/dbtest_neo4j.java
import java.io.*;
import java.util.*;

import org.neo4j.driver.v1.*;

/*

Random write: 247.56763 objects/second
Random read: 267.50842 objects/second

 */

public class dbtest_neo4j {
  private static final int NB = 10000;
  
  public static void main( String... args ) throws Exception {
    long t, dt;
    Driver db = GraphDatabase.driver( "bolt://localhost", AuthTokens.basic( "", "" ) );
    Session session = db.session();
    Transaction tx = session.beginTransaction();
    tx.run("CREATE INDEX ON :Node(label)");
    tx.success();
    tx.close();
    
    t = (new Date().getTime());
    tx = session.beginTransaction();
    Value node;
    Value previous = null;
    for (int i = 0; i < NB; i++) {
      String label = "node number " + i;
      node = tx.run("CREATE (x:Node {label:'" + label + "'}) RETURN id(x)").single().get(0);
      if (previous != null) {
        tx.run("MATCH (x:Node) WHERE id(x)=" + previous + " MATCH (y:Node) WHERE id(y)=" + node + " CREATE (x)-[:NEXT]->(y)");
      }
      previous = node;
    }
    tx.success();
    tx.close();
    dt = (new Date().getTime()) - t;
    System.err.print("Random write: "); System.err.print(1.0f/(dt/1000.0f)*NB); System.err.println(" objects/second");

    t = (new Date().getTime());
    tx = session.beginTransaction();
    node = tx.run("MATCH (x:Node) WHERE x.label='node number 0' RETURN id(x)").single().get(0);
    while (true) {
      Value label = tx.run("MATCH (x:Node) WHERE id(x)=" + node + " RETURN x.label").single().get(0);
      try {
        node = tx.run("MATCH (x:Node)-[:NEXT]->(y:Node) WHERE id(x)=" + node + " RETURN id(y)").single().get(0);
      }
      catch (org.neo4j.driver.v1.exceptions.NoSuchRecordException e) { break; }
    }
    tx.success();
    tx.close();
    dt = (new Date().getTime()) - t;
    System.err.print("Random read: "); System.err.print(1.0f/(dt/1000.0f)*NB); System.err.println(" objects/second");
    
  }
}

/*

javac -cp ~/src/neo4j-java-driver-1.0.0.jar ./dbtest_neo4j.java 

java -cp ~/src/neo4j-java-driver-1.0.0.jar:. dbtest_neo4j


 Supprime toute la base :

MATCH (n)  OPTIONAL MATCH (n)-[r]-()  DELETE n,r

    

*/
